core/src/main/java/org/apache/accumulo/core/client/security/tokens/PasswordToken.java

```diff
@@ -83,12 +86,46 @@ public PasswordToken(ByteBuffer password) {
 
   @Override
   public void readFields(DataInput arg0) throws IOException {
-    password = WritableUtils.readCompressedByteArray(arg0);
+    int version = arg0.readInt();
+    // -1 is null, consistent with legacy format; legacy format length must be >= -1
+    // so, use -2 as a magic number to indicate the new format
+    if (version == -1) {
+      password = null;
+    } else if (version == -2) {
+      byte[] passwordTmp = new byte[arg0.readInt()];
+      arg0.readFully(passwordTmp);
+      password = passwordTmp;
+    } else {
+      // legacy format; should avoid reading/writing compressed byte arrays using WritableUtils,
+      // because GZip is expensive and it doesn't actually compress passwords very well
+      AtomicBoolean calledFirstReadInt = new AtomicBoolean(false);
+      DataInput wrapped = (DataInput) Proxy.newProxyInstance(DataInput.class.getClassLoader(),
+          arg0.getClass().getInterfaces(), (obj, method, args) -> {
+            // wrap the original DataInput in order to return the integer that was read
+            // and then not used, because it didn't match -2
+            if (!calledFirstReadInt.get() && method.getName().equals("readInt")) {
+              calledFirstReadInt.set(true);
+              return version;
+            }
+            try {
+              return method.invoke(arg0, args);
+            } catch (InvocationTargetException e) {
+              throw e.getCause();
+            }
+          });
+      password = WritableUtils.readCompressedByteArray(wrapped);
+    }
```

Comment:
I am not sure what you are doing here with the legacy format.

Reply:
The original version just passed the full DataInput content to `WritableUtils.readCompressedByteArray()`. However, since I've already read the first integer from the front of the byte array, I need to inject it back into the front of the DataInput. I did that using Java's reflection Proxy to wrap the original DataInput. The first time `readInt` is called on the proxy, it provides the integer that was already previously pulled out of the original DataInput. After that, the proxy just passes through to the original DataInput.